components/uniset/src/props.rs

```diff
@@ -0,0 +1,402 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/master/LICENSE ).
+#![allow(clippy::unreadable_literal, dead_code)]
+
+use crate::provider::*;
+use crate::{UnicodeSet, UnicodeSetError};
+use icu_provider::prelude::*;
+use std::borrow::Cow;
+use std::convert::TryInto;
+
+// helper fn
+fn get_prop<'d, D: DataProvider<'d, UnicodeProperty<'d>> + ?Sized>(
+    ppucd_provider: &D,
+    resc_key: ResourceKey,
+) -> Result<UnicodeSet, UnicodeSetError> {
+    let data_req = DataRequest {
+        resource_path: ResourcePath {
+            key: resc_key,
+            options: ResourceOptions {
+                variant: None,
+                langid: None,
+            },
+        },
+    };
+    let mut resp: DataResponse<UnicodeProperty> = ppucd_provider.load_payload(&data_req)?;
+
+    let ppucd_property_cow: Cow<UnicodeProperty> = resp.take_payload()?;
+    let ppucd_property: UnicodeProperty = ppucd_property_cow.into_owned();
+    ppucd_property.try_into()
+}
+
+/// Upper
+pub fn upper<'d, D: DataProvider<'d, UnicodeProperty<'d>> + ?Sized>(
+    provider: &D,
+) -> Result<UnicodeSet, UnicodeSetError> {
+    get_prop(provider, key::UPPER_V1)
+}
```

Comment:
It seems like a great opportunity for a macro to minimize code redundancy and simplify maintenance.

Reply:
Yes and no.  I think that we should write fuller documentation for each of these functions, and the documentation will be harder to put in a macro.

Reply:
> Yes and no. I think that we should write fuller documentation for each of these functions, and the documentation will be harder to put in a macro.

```rust
/// Here is my doc for a function A
/// and I believe it will work.
func_prop(key::UPPER_V1);

// Here is my doc for a function B
func_prop(key::ANOTHER_KEY);
```

That wouldn't work?

Reply:
I haven't seen that construction, but if rustdoc is smart enough to expand the `func_prop!`, this sounds great.